java/runtime/src/main/java/io/ray/runtime/exception/NativeRayException.java

```diff
@@ -0,0 +1,214 @@
+
+    private final PrintStream printStream;
+
+    WrappedPrintStream(PrintStream printStream) {
+      this.printStream = printStream;
+    }
+
+    Object lock() {
+      return printStream;
+    }
+
+    void println(Object o) {
+      printStream.println(o);
+    }
+  }
+
+  private static class WrappedPrintWriter extends NativeRayException.PrintStreamOrWriter {
+
+    private final PrintWriter printWriter;
+
+    WrappedPrintWriter(PrintWriter printWriter) {
+      this.printWriter = printWriter;
+    }
+
+    Object lock() {
+      return printWriter;
+    }
+
+    void println(Object o) {
+      printWriter.println(o);
+    }
+  }
+
+  public NativeRayException(RayRuntimeInternal runtime, ErrorType errorType, Throwable e,
+      NativeRayException cause) {
+    RuntimeContext runtimeContext = runtime.getRuntimeContext();
+    WorkerContext workerContext = runtime.getWorkerContext();
+    StackTraceElement[] stackTrace = filterTrace(e);
+    String file = stackTrace[0].getFileName();
+    int lineNo = stackTrace[0].getLineNumber();
+    String function = stackTrace[0].getMethodName();
+    String traceBack = traceToString(stackTrace);
+    Pair<byte[], Boolean> serialized = Serializer.encode(e);
+    this.nativeHandle = nativeCreateRayException(
+        errorType.getNumber(),
+        cause == null ? e.getMessage() : "Get object failed because of deeper errors.",
+        Language.JAVA_VALUE,
+        runtimeContext.getCurrentJobId().getBytes(),
+        workerContext.getCurrentWorkerId().getBytes(),
+        workerContext.getCurrentTaskId().getBytes(),
+        workerContext.getCurrentActorId().getBytes(),
+        ObjectId.nil().getBytes(),
+        runtime.getRayConfig().nodeIp,
+        SystemUtil.pid(),
+        "Java Worker",
+        file,
+        lineNo,
+        function,
+        traceBack,
+        serialized.getLeft(),
+        cause == null ? new byte[0] : cause.toBytes());
+  }
+
+  protected NativeRayException(byte[] serialized) {
+    this.nativeHandle = nativeDeserialize(serialized);
+  }
+
+  public static NativeRayException fromBytes(byte[] serialized) {
+    return new NativeRayException(serialized);
+  }
+
+  public byte[] toBytes() {
+    return nativeSerialize(this.nativeHandle);
+  }
+
+  @Override
+  public void printStackTrace(PrintStream s) {
+    printStackTrace(new NativeRayException.WrappedPrintStream(s));
+  }
+
+  @Override
+  public void printStackTrace(PrintWriter s) {
+    printStackTrace(new NativeRayException.WrappedPrintWriter(s));
+  }
+
+  private void printStackTrace(NativeRayException.PrintStreamOrWriter s) {
+    synchronized (s.lock()) {
+      String stack = traceToString(filterTrace(this));
+      s.println(stack);
+      s.println("Caused by:");
+      s.println(this);
+    }
+  }
+
+  @Override
+  public String toString() {
+    return "\n" + nativeToString(this.nativeHandle);
+  }
+
+  @Override
+  protected void finalize() {
```

Comment:
`finalize` is deprecated in java9 and not recommended in java because it affects performance and gc . I think we can use java `phantom reference` and reference queue for this.

Reply:
You can take `io.ray.streaming.runtime.transfer.ChannelID#from(java.lang.String, byte[])` as an example

Reply:
I fixed this problem by using `phantom reference` instead of `finalize`.